File: spec/ruby/run_ext2.rb
#!/usr/bin/env ruby

require 'benchmark'

require 'ext2'

ITERATIONS = 30

def calc_dir(dir)
  sum = 0
  dir.entries.each { |entry|
    next if entry.name == '.' or entry.name == '..'
    case entry.file_type
    when :file_type_dir
      sum += calc_dir(entry.inode.as_dir)
    when :file_type_reg_file
      sum += entry.inode.size
    end
  }
  sum
end

t1_sum = 0
t2_sum = 0
fn = "#{ENV['DATA_DIR']}/ext2.dat"

ITERATIONS.times { |i|
  r = nil

  t1 = Benchmark::measure {
    r = Ext2::from_file(fn)
  }

  t2 = Benchmark::measure {
    p calc_dir(r.root_dir)
  }

  t1_sum += t1.real
  t2_sum += t2.real
}

puts "load: #{t1_sum.to_f / ITERATIONS}"
puts "calc: #{t2_sum.to_f / ITERATIONS}"

File: spec/python/run_ext2.py
#!/usr/bin/env python

import os
from timeit import default_timer as timer
from ext2 import Ext2

t1 = timer()

r = Ext2.from_file("%s/ext2.dat" % os.environ['DATA_DIR'])

t2 = timer()

def calc_dir(dir):
    sum = 0
    for entry in dir.entries:
        if entry.name == '.' or entry.name == '..':
            continue
        if entry.file_type == Ext2.DirEntry.FileType.dir:
            sum += calc_dir(entry.inode.as_dir)
        elif entry.file_type == Ext2.DirEntry.FileType.reg_file:
            sum += entry.inode.size
    return sum

calc_dir(r.root_dir)

t3 = timer()

print("parsing: %s seconds" % (t2-t1,))
print("processing: %s seconds" % (t3-t2,))
